### DISCUSSION TITLE
Incorporate Pupil Core's 2D pupil detectors as an algorithm

### DISCUSSION THREAD

**Discussion by cancan101** [openPupil/Open-PupilEXT#46] [Ideas]
Add another algorithm to the library from Pupil Core.

Implementation: https://github.com/pupil-labs/pupil-detectors
Paper (I think): https://dl.acm.org/doi/10.1145/2638728.2641695
Approach:
 1) Eye image converted to grayscale, user region of interest (white stroke rectangle), and initial estimation of pupil region (white square and dashed line square.) 
2) Canny edge detection (green lines.) 
3) Define “dark” region as offset from lowest spike in histogram within eye image. 
4) Filter edges to exclude spectral reflections (yellow.) and not inside “dark” areas (blue) 
5) Remaining edges extracted into contours using connected components and split into sub-contours based on curvature continuity criteria (multi colored lines). 
6) Candidate pupil ellipses (blue) are formed using ellipse fitting 
7) Final ellipse fit found through an augmented combinatorial search ( finally ellipse with center in red) - supporting edge pixels drawn in white.

**Comment by BZandi** [comment]
If you like, you can create a pull request and add this algorithm to PupilEXT. However, please check the licensing terms of the Pupil Core algorithm to determine whether this is possible or not. We provide a community branch for PupilEXT, allowing to add new features via pull requests (see here: https://github.com/openPupil/Open-PupilEXT/tree/Experimental-Community-Version)

**Comment by kheki4** [comment]
Hi there @cancan101 . Me and a friend of mine @sasfog are keen on adding new features to the application. We are working on the branch Babak also mentioned. We are currently finishing a new set of features, and testing stability. Once we are done with that, we will prepare a new release in the summer. After that, we will head to new directions: try to add entry-level gaze detection, and possibly try new pupil detection algorithms too. But it is too soon to say dates or exact priorities for now. Thanks for your idea! Please stay tuned for updates :)
